libraries/Microsoft.Bot.Builder.Core.Extensions/TranscriptLoggerMiddleware.cs

```diff
@@ -0,0 +1,126 @@
+                    context.Activity.From.Properties["role"] = "user";
+
+                LogActivity(CloneActivity(context.Activity));
+            }
+
+            // hook up onSend pipeline
+            context.OnSendActivities(async (ctx, activities, nextSend) =>
+            {
+                // run full pipeline
+                var responses = await nextSend();
+
+                foreach (var activity in activities)
+                {
+                    LogActivity(CloneActivity(activity));
+                }
+                return responses;
+            });
+
+            // hook up update activity pipeline
+            context.OnUpdateActivity(async (ctx, activity, nextUpdate) =>
+            {
+                // run full pipeline
+                var response = await nextUpdate();
+
+                // add Message Update activity
+                var updateActivity = CloneActivity(activity);
+                updateActivity.Type = ActivityTypes.MessageUpdate;
+                LogActivity(updateActivity);
+                return response;
+            });
+
+            // hook up delete activity pipeline
+            context.OnDeleteActivity(async (ctx, reference, nextDelete) =>
+            {
+                // run full pipeline
+                await nextDelete();
+
+                // add MessageDelete activity
+                var deleteActivity = new Activity() { Type = ActivityTypes.MessageDelete }.AsMessageDeleteActivity();
+                deleteActivity.Id = reference.ActivityId;
+                deleteActivity.From = reference.Bot;
+                deleteActivity.Conversation = reference.Conversation;
+                deleteActivity.Recipient = reference.User;
+                deleteActivity.ChannelId = reference.ChannelId;
+                deleteActivity.Timestamp = DateTimeOffset.UtcNow;
+
+                LogActivity(deleteActivity);
+            });
+
+            // process bot logic
+            await nextTurn().ConfigureAwait(false);
+
+            // flush transcript at end of turn
+            while (transcript.Count > 0)
+            {
+                try
+                {
+                    var activity = transcript.Dequeue();
+                    await logger.LogActivity(activity);
+                }
+                catch (Exception err)
+                {
+                    System.Diagnostics.Trace.TraceError($"Transcript logActivity failed with {err}");
+                }
+            }
+        }
+
+        private void LogActivity(IActivity activity)
+        {
+            lock (transcript)
+            {
+                if (activity.Timestamp == null)
+                    activity.Timestamp = DateTime.UtcNow;
+
+                transcript.Enqueue(activity);
+            }
+        }
+
+        private static IActivity CloneActivity(IActivity activity)
+        {
+            activity = JsonConvert.DeserializeObject<Activity>(JsonConvert.SerializeObject(activity, JsonSettings));
```

Comment:
consider defining a local variable here, rather than re-using the argument passed in. 
IActivity cloned = [...]